packages/core-db/src/app/queued-persisted-processor.ts

```diff
@@ -0,0 +1,330 @@
+   *
+   * @param index The index in question.
+   * @returns The fetched item if it exists, undefined otherwise.
+   */
+  private async fetchItem(index: number): Promise<T | undefined> {
+    const itemBuffer: Buffer = await this.db.get(
+      this.getStorageKeyForIndex(index)
+    )
+    if (!itemBuffer) {
+      return undefined
+    }
+    return this.deserializeItem(itemBuffer)
+  }
+
+  /**
+   * Gets the storage key for the item with the provided index.
+   *
+   * @param index The index in question.
+   * @returns The storage key (Buffer) of the item in question.
+   */
+  private getStorageKeyForIndex(index: number): Buffer {
+    return this.getStorageKey(
+      `${BaseQueuedPersistedProcessor.ITEM_STORAGE_KEY_PREFIX}${index}`
+    )
+  }
+
+  /**
+   * Gets this instance's storage key for the item with the provided key.
+   * This takes into account the fact that there may be multiple instances of this class
+   * running at once.
+   *
+   * @param key The key in question.
+   * @returns The storage key (Buffer) of the item in question.
+   */
+  private getStorageKey(key: string): Buffer {
+    return Buffer.from(`${this.persistenceKey}_${key}`)
+  }
+
+  /**
+   * Sets the last processed index, persisting the updated index in case of failure.
+   * @param index The index to set Last Processed to.
+   */
+  private async setLastProcessed(index: number): Promise<void> {
+    await this.db.put(
+      this.getStorageKey(BaseQueuedPersistedProcessor.LAST_INDEX_PROCESSED),
+      BaseQueuedPersistedProcessor.serializeNumber(index)
+    )
+    this.lastIndexProcessed = index
+    this.log(`Last processed incremented to ${this.lastIndexProcessed}`)
+  }
+
+  /**
+   * Sets the next index to process, persisting the updated index in case of failure.
+   * @param index The index to set  Processed to.
+   */
+  private async setNextToProcess(index: number): Promise<void> {
+    await this.db.put(
+      this.getStorageKey(
+        BaseQueuedPersistedProcessor.NEXT_INDEX_TO_PROCESS_KEY
+      ),
+      BaseQueuedPersistedProcessor.serializeNumber(index)
+    )
+    this.nextIndexToProcess = index
+    this.log(`Next to process incremented to ${this.nextIndexToProcess}`)
+  }
+
+  /**
+   * Utility to make number serialization consistent.
+   * This must be compatible with deserializeNumber(...).
+   *
+   * @param num The number to be serialized.
+   * @returns The serialized number as a Buffer.
+   */
+  private static serializeNumber(num: number): Buffer {
+    const hex: string = num.toString(16)
+    return hexStrToBuf(hex.length % 2 === 0 ? hex : `0${hex}`)
+  }
+
+  /**
+   * Utility to make number deserialization consistent.
+   * This must be compatible with serializeNumber(...).
+   *
+   * @param buf The buffer to deserialize into a number
+   * @returns The deserialized number.
+   */
+  private static deserializeNumber(buf: Buffer): number {
+    return parseInt(bufToHexString(buf, false), 16)
+  }
+}
```

Comment:
Yeah. Theres `bufferUtils.numberToBuffer(...)` but not the reverse, and if we add one, it may not be clear or enforced that they _must_ be compatible with each other.